feat(navigation): add Navigation.Vertical component

bryceosterhaus opened the pull request:
Wanted to send this to get external opinions. As of now, I sort of see `Vertical` as a high-level component. I also created low-level nav components but I wasn't totally sure how we want to expose those within the navigation package.

Let me know your thoughts.

TODO:
- [x] exporting nav components
- [x] tests
- [x] docs

wincent commented on packages/clay-navigation/src/Vertical.tsx:
```diff
@@ -0,0 +1,128 @@
+	level = 0
+) {
+	return items.map(({active, href, items: subItems, label}, i) => {
+		const key = `${level}${i}`;
```
Shouldn't we have a separator here? Absurd example: `level` 1/`i` 11 and `level` 11/ `i` 1 produce the same key (111). May not actually have a collision in practice but in general, if you are making a compound key out of two numbers, you need a separator to isolate the components.

bryceosterhaus replied:
Good catch. just pushed an update.

matuzalemsteles commented on packages/clay-navigation/src/NavLink.tsx:
```diff
@@ -0,0 +1,59 @@
+	...otherProps
+}) => {
+	return (
+		<a
```
Could we use `ClayLink` here?

matuzalemsteles commented on packages/clay-navigation/src/Vertical.tsx:
```diff
@@ -0,0 +1,128 @@
+
+		return (
+			<NavItem key={i}>
+				<NavLink
```
Chances are we also need to use a `ClayButton` here for when people want to control the component instead of creating link navigation.

bryceosterhaus commented:
@wincent @matuzalemsteles  updated.

I added ClayLink. We might want to look into using `ClayLinkOrButton` in the future, but as of not the css doesn't seem to work well for it.

I also removed the internal map entirely and rely on internal state instead. This allows for cleaner logic and gives easier use of animation. I don't see this being an issue, since we still allow users to set the initial expanded value in `items`.

matuzalemsteles commented on packages/clay-navigation/package.json:
```diff
@@ -26,6 +26,7 @@
 		"react"
 	],
 	"dependencies": {
```
Remember to add `@clayui/icon` here

matuzalemsteles commented on packages/clay-navigation/src/Vertical.tsx:
```diff
@@ -0,0 +1,143 @@
+				['menubar-vertical-expand-md']: !large,
+			})}
+		>
+			<button
```
Just to know, could we use a `ClayButtonWithIcon` here?

bryceosterhaus replied:
i forgot about that one. good catch

bryceosterhaus replied:
Actually, I just realized we need to use it without `ClayButtonWithIcon` since this component requires `{activeLabel}` to be in the button and next to the icon. `ClayButtonWithIcon` is only for just an icon.

matuzalemsteles commented:
> I also created low-level nav components but I wasn't totally sure how we want to expose those within the navigation package.

I think it makes sense to expose as a low-level component too since people can compose `NavLink` with other elements or allow people to customize.

Can we expose packages with the same idea we did for `ClayForm` to shorten namespaces?

bryceosterhaus commented:
@matuzalemsteles  updated

matuzalemsteles commented on packages/clay-navigation/src/index.tsx:
```diff
@@ -6,13 +6,25 @@
+ClayNavigation.Nav = Nav;
+ClayNavigation.NavItem = NavItem;
+ClayNavigation.NavLink = NavLink;
+ClayNavigation.Vertical = Vertical;
```
I was thinking of making it available the same way we did in `ClayForm`. It seems a little strange... what do you think?

```javascript
ClayVertical.Nav = Nav;
ClayVertical.NavItem = NavItem;
ClayVertical.NavLink = NavLink;

export {ClayVertical};
```

Now I've been wondering if all the components should be in the same namespace, maybe you get the impression that it's all part of `Navigation`, `Form` and you can compose them all anywhere, it looks like we can sell more...

bryceosterhaus replied:
I didn't want to namespace it off of `ClayVertical` because `nav` is used outside of the context of `Vertical` in other types of navs.

We could possible change it from `@clayui/navigation` to `@clayui/nav` and the way we can do something like

```js
ClayNav = Nav;
ClayNav.Item = NavItem;
ClayNav.Link = NavLink;


// option #1
ClayNav.Vertical = Vertical;
ClayNav.BreadCumb = BreadCumb;

// option #2
export {
       ClayNavWithBreadCrumb,
       ClayNavWithVertical,
}

export default ClayNav
```

matuzalemsteles replied:
Hmm, makes sense. I think we can keep this within `Navigation` even though we may also migrate `Navigation Bar` to it.

bryceosterhaus replied:
Right on, so ship components like so? This means our root offering for Navigation is `'nav'`. 

```js
ClayNavigation = Nav;
ClayNavigation.Item = NavItem;
ClayNavigation.Link = NavLink;
ClayNavigation.Vertical = Vertical;
ClayNavigation.BreadCumb = BreadCumb;

export default ClayNavigation
```

This makes sense to me I think.

matuzalemsteles replied:
Hmm, maybe so. Not sure about this because the breadcrumb base is not a nav but I think we can go with that.

bryceosterhaus commented:
@matuzalemsteles updated and added docs and tests.

The export structure I went with is

```js
import Nav from './Nav';
import {ClayBreadcrumbNav} from './Breadcrumb';
import {ClayVerticalNav} from './Vertical';

export {ClayBreadcrumbNav, ClayVerticalNav};

export default Nav;
```

I went with named exports because if I namespaces `Vertical` on `Nav`, it would create a circular dependency between the two.
